Add hooks before saving resource

yggie opened the pull request:
Partly addresses https://github.com/apokalipto/devise_saml_authenticatable/issues/38

This PR introduces a mechanism to apply custom behaviour prior to saving
a resource. This behaviour occurs when `Devise.saml_update_user` or
`Devise.saml_create_user` has been enabled, and the resource has been
found or created. The mechanism works by means of introducing a hook
just prior to saving the resource, providing an application the
flexibility to make arbitrary changes to the resource.

This hook replaces the existing `set_user_saml_attributes` with a
default hook implementation which can be replaced or extended depending
on the application needs.

The changes are backward compatible to my knowledge, public
configuration has not been modified, but a new configuration has been
introduced (`saml_update_resource_hook`).

yggie commented:
If there are any suggestions on how this can be improved, please do not hesitate to suggest. This solves my own problem I was having with the library (having some hook to update the user record), but it might not fit everyones needs. If the approach is all good, I can spend some time adding documentation and whatnot before we merge it 😄 

PS: great library!

adamstegman commented on lib/devise_saml_authenticatable/model.rb:
```diff
@@ -52,7 +55,7 @@ def authenticate_with_saml(saml_response, relay_state)
+            Devise.saml_update_resource_hook.call(resource, decorated_response)
             if (Devise.saml_use_subject)
               resource.send "#{key}=", auth_value
             end
```
Seems like this (and the following `save!` call) should go within the block as well.

adamstegman commented on lib/devise_saml_authenticatable/saml_response.rb:
```diff
@@ -0,0 +1,16 @@
+require 'devise_saml_authenticatable/saml_mapped_attributes'
+
+module SamlAuthenticatable
+  class SamlResponse
```
Is this class necessary? It seems like a simple factory for `SamlMappedAttributes`, which doesn't seem like it's needed since the constructor is so straightforward.

adamstegman commented on lib/devise_saml_authenticatable/saml_mapped_attributes.rb:
```diff
@@ -0,0 +1,25 @@
+      @inverted_attribute_map = @attribute_map.invert
+    end
+
+    def saml_attribute_keys()
```
Can you remove the empty parentheses? ✨

yggie commented:
apologies this completely slipped my mind! if https://github.com/apokalipto/devise_saml_authenticatable/pull/81 addresses the comments, I will be happy to have that merged and close this

ph0rque commented:
Can you give an example of how to use the hooks? I need to create an auto-generated random password for a new user, set the registration type to `saml`, etc.
